R/join_ald_scenario.R

```diff
@@ -48,14 +48,93 @@ join_ald_scenario <- function(data,
   ald <- modify_at_(ald, "technology", tolower)
   out <- data %>%
     left_join(ald, by = ald_columns()) %>%
-    inner_join(scenario, by = scenario_columns()) %>%
-    warn_if_has_zero_rows("Joining `scenario` outputs 0 rows.") %>%
-    mutate(plant_location = tolower(.data$plant_location)) %>%
-    inner_join(
-      region_isos,
-      by = c("region", "plant_location" = "isos", "scenario_source")
-    ) %>%
+    mutate(plant_location = tolower(plant_location)) %>%
+    left_join(region_isos, by = c("plant_location" = "isos")) %>%
     warn_if_has_zero_rows("Joining `region_isos` outputs 0 rows.")
+  all_company <- unique(out$name_ald)
+  all_region <- unique(out$region)
+  for (one_region in all_region) {
+    # print(one_region)
+    for (company in all_company) {
+      filtered_data <- out %>%
+        filter(
+          name_ald == company
+        )
+      sector_company <- filtered_data$sector_ald[1]
+      all_tech <- green_or_brown %>%
+        filter(sector == sector_company)
+      all_tech <- all_tech$technology
+      green_tech <- green_or_brown %>%
+        filter(
+          sector == sector_company,
+          green_or_brown == "green"
+        )
```

Comment:
Yes I wasn't really sure how I can easily add a lot of lines giving some filter, for now that is the only but not ideal way I found (combining `for` and `rbind` based on some filters) .
`right_join` may be a great trick I will look at it tomorrow

I don't understand your last point about this `all_technologies` parameters. The default parameters would output something that is not our methodology or neither accorded to scenario. But I understand it might also be a problem to change significatly the `join_ald_scenario` function.

Reply:
@Antoine-Lalechere the second issue is coming from a software maintenance point of view. 
`r2dii.analysis` is a package that has been published on CRAN, and has a handful of "exported" functions. 
See here for an overview: https://2degreesinvesting.github.io/r2dii.analysis/reference/index.html

`join_ald_scenario` is one of these "exported" functions, which means we are restricted in how much we can change its behaviour. `join_ald_scenario` is expected to "Return a fully joined data frame, linking portfolio, ald and scenario.", but it is not necessarily expected to fill out technologies that are in `scenario`, but not the other input datasets. 

Changing how an exported function behaves (when using the exact same arguments that it used before) is generally a bad idea in software development, as you risk breaking any upstream pipelines that use the exported functions. 

One way around this is to instead add an argument. With the new arguments default value, the function should behave as previously expected, but changing that new argument to some other value, you can add a functionality to the exported function. 

I'm not sure if this is the appropriate decision in this case, but it's something we need to consider.

(I of course did not expect you to know any of these more software development/ maintenance-oriented points prior to submitting this PR, so absolutely no worries at all, just something we will have to consider and figure out the solution :-) )

Reply:
Sounds super interesting, I will go look a bit deeper into it - in that case yes it might be an option but I'm probably not the best to know all the pros and cons.

When I try to use `right_join` at a portfolio level, it adds the technology that are not in my loanbook (ex. nuclear in ald_demo) but with a full line of NA (i.e. no company name or loan informations) and at a company level, it adds lines for all sectors and all years but without information on the company. I will try to dig a bit into this and if not looking how we can make it compute best and quicker

Reply:
You will need to do a `group_by`, and then explicitly fill the NA's in the appropriate groups.

Reply:
If I want to only some join and to make appear a line where `name_ald=="dynegy ..." `and `technology=="renewablescap"`, I can't do it with a `right_join` or I should remove "technology" from `scenario_columns`. But in the second case it becomes a mess to then eliminate useless lines.

Reply:
In fact this function would also make the job and is far faster than the previous one, thanks for your advise @jdhoffa!

```
join_ald_scenario <- function(data,
                               ald,
                               scenario,
                               region_isos = r2dii.data::region_isos
                               ) {
  check_portfolio_ald_scenario(data, ald, scenario)
  
  # Track provenance to avoid clash in the column name "source"
  region_isos <- region_isos %>%
    rename(scenario_source = .data$source)
  
  crucial_groups <- c(
    "id_loan",
    "loan_size_outstanding",
    "loan_size_outstanding_currency",
    "loan_size_credit_limit",
    "loan_size_credit_limit_currency",
    "name_ald",
    "sector_ald",
    "technology",
    "year",
    "scenario",
    "region",
    "tmsr",
    "smsp",
    "scenario_source"
  )
  
  ald <- modify_at_(ald, "sector", tolower)
  ald <- modify_at_(ald, "technology", tolower)
  out <- data %>%
    left_join(ald, by = ald_columns()) %>%
    group_by(.data$name_ald)%>%
    mutate(plant_location = tolower(.data$plant_location))

  out<-out%>%
    right_join(scenario, by = c(scenario_columns())) %>%
    mutate(
      production=ifelse(technology.x==technology.y,production,0)
    )%>%
    select(-technology.x)%>%
    rename(technology=technology.y)%>%
    warn_if_has_zero_rows("Joining `scenario` outputs 0 rows.") %>%
    inner_join(
      region_isos,
      by = c("region", "plant_location" = "isos", "scenario_source")
    ) %>%
    warn_if_has_zero_rows("Joining `region_isos` outputs 0 rows.")%>%
    select(-country_of_domicile, -plant_location)%>%
    distinct()%>%
    group_by(!!!rlang::syms(crucial_groups))%>%
    summarize(
      production = sum(.data$production))%>%
    filter( 
        toString(production)!="NA")
  out
}
```